### PR TITLE
sort constraints of hpolygon on construction

### LINKED ISSUE #477: Constructors of AbstractHPolygon with sorting
See #475.

### LINKED ISSUE #736: HPolygon conversions need to sort the constraints
Some `convert` methods to `HPolygon` do not sort the constraints, e.g., [this one](https://github.com/JuliaReach/LazySets.jl/blob/372c19ae66309d740769f1d60b35aff30496fac4/src/convert.jl#L357).

### PR DESCRIPTION
**Pull request by mforets**
Closes #477.
Closes #736.

### CODE REVIEW

File: src/HPolygonOpt.jl [hunk 1]
```diff
@@ -34,10 +34,20 @@ mutable struct HPolygonOpt{N<:Real} <: AbstractHPolygon{N}
...
-    # default constructor
-    HPolygonOpt{N}(constraints::Vector{LinearConstraint{N}},
-                   ind::Int=1) where {N<:Real} =
-        new{N}(constraints, ind)
+    # default constructor that applies sorting of the given constraints
+    function HPolygonOpt{N}(constraints::Vector{LinearConstraint{N}},
+                            ind::Int=1;
+                            sort_constraints::Bool=true) where {N<:Real}
+        if sort_constraints
+            sorted_constraints = Vector{LinearConstraint{N}}()
```
Comment (schillic) [inline src/HPolygonOpt.jl]: `sizehint!`

File: src/HPolygon.jl [hunk 2]
```diff
@@ -26,8 +26,27 @@ Use `addconstraint!` to iteratively add the edges in a sorted way.
 """
 struct HPolygon{N<:Real} <: AbstractHPolygon{N}
     constraints::Vector{LinearConstraint{N}}
+
+    # default constructor that applies sorting of the given constraints
+    function HPolygon{N}(constraints::Vector{LinearConstraint{N}};
+                         sort_constraints::Bool=true) where {N<:Real}
+        if sort_constraints
+            sorted_constraints = Vector{LinearConstraint{N}}()
```
Comment (schillic) [inline src/HPolygon.jl]: `sizehint!`

File: src/AbstractHPolygon.jl [hunk 3]
```diff
@@ -206,29 +204,53 @@ Nothing.
...
+    return addconstraint!(P.constraints, constraint, linear_search=linear_search)
+end
+
+"""
+    addconstraint!(constraints::Vector{LinearConstraint{N}},
+                   new_constraint::LinearConstraint{N};
+                   [linear_search]::Bool=(length(P.constraints) < BINARY_SEARCH_THRESHOLD)
+                  )::Nothing where {N<:Real}
+
+Add a linear constraint to a polygon in constraint representation, keeping the
```
Comment (schillic) [inline src/AbstractHPolygon.jl]: Needs a change - something along the line of
```
Add a linear constraint to a sorted vector of constraints
```

**Comment by schillic** [comment]
I think this closes #477 and is very related to #736.

**Comment by mforets** [comment]
> I think this closes #477 and is very related to #736.

ok, i think it closes both because `HPOLYGON{N}(constraints_list(H))` will sort the constraints in the construction

**Comment by mforets** [comment]
> Sorry, but we (= you) should check all places where an HPolygon is created and potentially pass sort_constraints=false if we already know that the constraints are sorted.

good catch! i found several and updated the branch.

**Pull request merged**